[ML] Change Point Detection

darnautov opened the pull request:
## Summary

Adds a Change point detection page in the AIOps labs.

_Note:_ 
This page will be hidden under the hardcoded feature flag for 8.6. 

<img width="1775" alt="image" src="https://user-images.githubusercontent.com/5236598/199506277-f0d71104-3098-4e15-a697-35f5eec5c110.png">

### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

darnautov commented:
@elasticmachine merge upstream

mdefazio reviewed (changes requested):
A few comments on the UI:

**Start screen:**

On the data view selection screen, I would recommend at least moving the title 'Select data view or saved search' inside the panel so the page is presented more like an empty state. And then show the app title in its place.  This is done on the File Data Visualizer page, but not on the other usages of this template. The breadcrumbs should show the page title, rather than the call-to-action to Select a data view.
<img width="558" alt="image" src="https://user-images.githubusercontent.com/3756330/199023185-75e761cd-0fe1-42d5-aeda-8ce051109771.png">
<img width="259" alt="image" src="https://user-images.githubusercontent.com/3756330/199023070-793ae4b3-d351-480d-81f4-f0c1b53bd02e.png">

- - - 

Should we include a data view dropdown similar to Discover? I have to go back to the main Change point detection page and select a new data view.  Would be nice to avoid this extra step and follow Discover's pattern.
<img width="1567" alt="image" src="https://user-images.githubusercontent.com/3756330/199019966-13d25d5a-ae5f-4e7c-9686-f54d0880eda6.png">

-  - -

Shouldn't the left nav be selected, even after I choose a data view? I noticed this also isn't the case in the other AIOps pages, as well as the Data visualizer pages.  However, the left nav menu item _is_ selected for the Anomaly Detection and Data Frame Analytics pages.  
<img width="642" alt="image" src="https://user-images.githubusercontent.com/3756330/199024974-c0d0693a-0cf1-4475-ad41-5e6a3610f2f8.png">
<img width="698" alt="image" src="https://user-images.githubusercontent.com/3756330/199025217-1898f981-78f9-4309-aca0-a6c612795cae.png">

- - - 

As for the layout of the page, I think the number of borders/dividers distracts from the graphs.  Perhaps we remove the divider underneath the dropdowns and put a large spacer in it's place.  I think we could also increase the gutter size to `medium` to visually separate each panel more.
<img width="1479" alt="image" src="https://user-images.githubusercontent.com/3756330/199026005-45245bcc-923b-4136-8cf0-e7f1cec07494.png">

- - - 

We could also consider condensing the top portion more by: 
* moving the dropdown labels as prepends, saving some height
* Using the dataview dropdown next to the search bar similar to Discover
* Likely also enough room to combine in the date picker / refresh buttons

Alternatively, maybe we can move the dropdowns (using prepend labels) into the same line as the search bar and keep the date picker as is?

- - -

Should the AIOps page show up in the global search?  Seems mixed as to which pages show and which don't.

- - -

I know some of this extends beyond this page.  But let me know if you need further explanations or wireframes/mocks for the suggestions.

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/chapge_point_detection_page.tsx:
```diff
@@ -0,0 +1,171 @@
+  );
+
+  return (
+    <div data-test-subj="aiopsChanePointDetectionPage">
```
Nit: Missing `g` in `aiopsChangePointDetectionPage`,

darnautov replied:
Fixed in https://github.com/elastic/kibana/pull/144093/commits/998cdd4777d337c22db72a311deacb070af587f8

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/chart_component.tsx:
```diff
@@ -0,0 +1,220 @@
+        {
+          type: 'index-pattern',
+          id: dataView.id!,
+          name: 'indexpattern-datasource-layer-2d61a885-abb0-4d4e-a5f9-c488caec3c22',
```
Can you explain where these names and ids in this specification come from?

darnautov replied:
I used the Lens playground to generate the attributes, you can find out more about how it's done in this guide https://github.com/elastic/kibana/tree/main/x-pack/plugins/lens#getting-started

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/chart_component.tsx:
```diff
@@ -0,0 +1,220 @@
+          ],
+        },
+        query: {
+          query: '',
```
Do we need to consider custom queries from the query bar here?

darnautov replied:
not sure it's needed, because the charts are rendered based on the aggregation results, and the change point agg query takes it into account

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/split_field_selector.tsx:
```diff
@@ -0,0 +1,37 @@
+        />
+      }
+    >
+      <EuiSelect options={options} value={value} onChange={(e) => onChange(e.target.value)} />
```
Suggestion maybe for a follow up: For datasets with a lot of attributes, it might be useful to use a selector that allows searching the fields.

darnautov replied:
It's a good point, I'm going to replace select with a [combo-box](https://elastic.github.io/eui/#/forms/combo-box)

peteharverson commented on x-pack/plugins/ml/public/application/routing/routes/new_job/index_or_search.tsx:
```diff
@@ -72,6 +72,17 @@ const getLogCategorizationBreadcrumbs = (navigateToPath: NavigateToPath, basePat
+const getChangePointDetectionBreadcrumbs = (navigateToPath: NavigateToPath, basePath: string) => [
+  getBreadcrumbWithUrlForApp('ML_BREADCRUMB', navigateToPath, basePath),
+  getBreadcrumbWithUrlForApp('AIOPS_BREADCRUMB_CHANGE_POINT_DETECTION', navigateToPath, basePath),
+  getBreadcrumbWithUrlForApp('CHANGE_POINT_DETECTION', navigateToPath, basePath),
```
If I click on the `AIOps Labs` breadcrumb here, as a way to select a new data view, the route switches to the Explain Log Rate Spikes page.

<img width="582" alt="image" src="https://user-images.githubusercontent.com/7405507/199224316-3a3a2b69-10f3-4c16-a914-9e3c49f20f3f.png">

darnautov replied:
fixed in https://github.com/elastic/kibana/pull/144093/commits/dbd1c5ef3f859f93c81b508fcb9bb57d88048352

peteharverson commented on x-pack/plugins/ml/public/application/routing/routes/new_job/index_or_search.tsx:
```diff
@@ -72,6 +72,17 @@ const getLogCategorizationBreadcrumbs = (navigateToPath: NavigateToPath, basePat
+  getBreadcrumbWithUrlForApp('AIOPS_BREADCRUMB_CHANGE_POINT_DETECTION', navigateToPath, basePath),
+  getBreadcrumbWithUrlForApp('CHANGE_POINT_DETECTION', navigateToPath, basePath),
+  {
+    text: i18n.translate('xpack.ml.aiopsBreadcrumbs.selectDateViewLabel', {
```
Typo in the key here - should be `selectDataViewLabel`

darnautov replied:
Fixed in https://github.com/elastic/kibana/pull/144093/commits/998cdd4777d337c22db72a311deacb070af587f8

peteharverson commented:
Adding a comment for my feedback - will add further feedback to this comment as I test.

- Should we filter out the `stationary` items, and instead show a 'No significant change points detected` if there are no significant deviations?

peteharverson commented on x-pack/plugins/aiops/public/components/change_point_detection/chapge_point_detection_page.tsx:
```diff
@@ -0,0 +1,171 @@
+                {v.p_value !== undefined ? (
+                  <EuiDescriptionList
+                    type="inline"
+                    listItems={[{ title: 'p_value', description: v.p_value }]}
```
The Log Rate Spikes page uses a hyphen in `p-value` rather than an underscore. Also the precision for the value should be reduced I think to be consistent with other pages, where 3 sigfigs is used:

<img width="207" alt="image" src="https://user-images.githubusercontent.com/7405507/199241941-af3209a2-6c2b-4043-8b12-b054df4517c8.png">

darnautov replied:
Updated in https://github.com/elastic/kibana/pull/144093/commits/e45965162a5a5775f1d07e9334e218e402068b04

peteharverson replied:
As discussed, I now see `NaN` for quite a lot of series:

<img width="353" alt="image" src="https://user-images.githubusercontent.com/7405507/201886989-636c33d2-ec42-4329-87ce-9fa073bb1abf.png">

darnautov replied:
fixed in https://github.com/elastic/kibana/pull/144093/commits/50979303ab2f771cc915490b3065d4a45fbb09a4

darnautov commented:
hi @mdefazio, thanks for the review! All the statements are valid, but for this PR I assume it's worth addressing the comment related to the new page directly. 

1. I fixed the active item side nav indicator for all AIOps pages 
2. Removed the divider underneath the controls and adjusted spacing based on your recommendations 
3. Replaced labels with `prepend`
4. There is an additional element needed on the page - a progress bar, you can spot it when dealing with the high cardinality data. I decided to put it in line with the controls to prevent the page from jumping when the progress bar is hidden. 

What do you think? 

<img width="1775" alt="image" src="https://user-images.githubusercontent.com/5236598/199507214-fab623ea-5103-454e-b680-2d1dbe35d9d6.png">

mdefazio reviewed (changes requested):
@darnautov Thanks for making those updates. 

I noticed we're using a callout to mention when No change points are found. I would suggest using an empty prompt here instead. 

<img width="755" alt="image" src="https://user-images.githubusercontent.com/3756330/199522513-d253ea97-c2d7-4602-80e5-fb9dcfd3d95d.png">


I think the progress bar placement works, my only suggestion would be to have `grow={false}` added to it—I don't think it needs to span the rest of the row. Though perhaps others have a differing opinion here—so not a deal breaker either way for me.

<img width="534" alt="image" src="https://user-images.githubusercontent.com/3756330/199528308-1ea754d1-50c0-4b6f-859e-bdab26ec41c6.png">

I noticed sometimes it went over 100% though...

<img width="350" alt="image" src="https://user-images.githubusercontent.com/3756330/199528105-4b976da3-5379-4bd7-8905-11cbcf166164.png">

Thanks again for making the edits! Just marking as change request for the empty prompt rather than a callout to be more consistent across our product.

darnautov commented:
thanks for checking @mdefazio, addressed your comments in the latest comments. Into callout has been replaced with an empty state, and the progress bar doesn't span the rest of the row anymore. 

<img width="1469" alt="image" src="https://user-images.githubusercontent.com/5236598/199713740-e7f1ad9e-e43b-470f-a29b-b564a4c26951.png">

mdefazio reviewed (approved):
Thanks for making those changes!

walterra commented:
I noticed we now have annotations for a point in time instead of the earlier range based one. It looks like it not completely aligned with the buckets of the date histogram aggregation, is there a way to adapt this?
 
<img width="596" alt="image" src="https://user-images.githubusercontent.com/230104/201085728-56fb41c9-2eff-40ff-839c-243e70d480c7.png">

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/change_point_detection_context.tsx:
```diff
@@ -0,0 +1,275 @@
+  );
+};
+
+export function useChangePontDetectionContext() {
```
Typo, should be `useChangePointDetectionContext`.

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/change_point_detection_context.tsx:
```diff
@@ -0,0 +1,275 @@
+}
+
+export function useRequestParams() {
+  return useChangePontDetectionContext().requestParams;
```
Typo, should be `useChangePointDetectionContext`.

darnautov replied:
Fixed in https://github.com/elastic/kibana/pull/144093/commits/92ce0632a18f52338f038b66937d45e2fa7facf6

walterra commented on x-pack/plugins/aiops/public/components/change_point_detection/function_picker.tsx:
```diff
@@ -0,0 +1,47 @@
+  return (
+    <EuiFormRow>
+      <EuiSelect
+        options={[
+          {
+            value: 'min',
+            text: 'min',
+          },
+          {
+            value: 'max',
+            text: 'max',
+          },
+          {
+            value: 'sum',
+            text: 'sum',
+          },
+          {
+            value: 'avg',
+            text: 'avg',
+          },
+        ]}
```
Nit: Maybe these options could be derived from `fnOperationTypeMapping` instead of being hard-coded again here?

darnautov commented:
> I noticed we now have annotations for a point in time instead of the earlier range based one. It looks like it not completely aligned with the buckets of the date histogram aggregation, is there a way to adapt this?
> 
> <img alt="image" width="596" src="https://user-images.githubusercontent.com/230104/201085728-56fb41c9-2eff-40ff-839c-243e70d480c7.png">

@walterra  I noticed it only happens with certain datasets 🤔 Any idea what might be the reason? I set the same interval for charts and the date histogram agg with change point detection

darnautov commented on x-pack/plugins/aiops/public/components/change_point_detection/chart_component.tsx:
```diff
@@ -0,0 +1,214 @@
+  };
+}
+
+export const ChartComponent: FC<ChartComponentProps> = React.memo(({ annotation }) => {
```
@flash1293 Could you please review this file and verify if usage of the Lens embeddable is correct? Especially the attributes part

peteharverson commented on x-pack/plugins/aiops/public/components/change_point_detection/use_change_point_agg_request.ts:
```diff
@@ -0,0 +1,271 @@
+
+  const splitFieldCardinality = useSplitFieldCardinality(requestParams.splitField, query);
+
+  const { runRequest, cancelRequest, isLoading } = useCancellableRequest();
```
Is the request cancelling after leaving the page. With a large data set (14.5M docs, split field cardinality 7700) I still see activity in the network tab after returning to the Overview page.

peteharverson reviewed (approved):
This looks great for an initial iteration. Would be great to get this in, but hidden for 8.6 behind a hardcoded feature flag. Please create a separate issue for the feedback in https://github.com/elastic/kibana/pull/144093#issuecomment-1298403320 and address as appropriate for 8.7.0.

jgowdyelastic commented on x-pack/plugins/aiops/public/components/change_point_detection/use_change_point_agg_request.ts:
```diff
@@ -0,0 +1,270 @@
+        cancelRequest();
+      };
+    },
+    [query, cancelRequest, requestParams]
```
it appears `cancelRequest` will be called more than once as the `useEffect` on line 219 is also watching `query` and `requestParams` and it calls `reset`.
Is there a risk of a race condition between these useEffects and the other trigger for `fetchResults`?

darnautov replied:
fixed in https://github.com/elastic/kibana/pull/144093/commits/1e0bdabdea409722d3c195f139b14f0c011392e9

jgowdyelastic commented on x-pack/plugins/aiops/public/components/change_point_detection/use_change_point_agg_request.ts:
```diff
@@ -0,0 +1,270 @@
+    [runRequest, requestParams, query, dataView, splitFieldCardinality, toasts]
+  );
+
+  useEffect(() => {
```
can this function be given a name?

jgowdyelastic commented on x-pack/plugins/aiops/public/hooks/use_cancellable_request.ts:
```diff
@@ -0,0 +1,67 @@
+import { tap } from 'rxjs/operators';
+import { useAiopsAppContext } from './use_aiops_app_context';
+
+export function useCancellableRequest() {
```
this function name isn't obvious that it's running a search. 
`useCancellableSearch` might be clearer?

darnautov replied:
renamed in https://github.com/elastic/kibana/pull/144093/commits/b068cebdbf7b21b36bb907bfe7f3d2df47b0cba0

jgowdyelastic commented on x-pack/plugins/aiops/public/hooks/use_data_source.ts:
```diff
@@ -0,0 +1,26 @@
+  dataView: DataView | never;
+  savedSearch: SavedSearch | SavedSearchSavedObject | null;
+}>({
+  // @ts-ignore
```
This ignore could do with a reason why

darnautov replied:
fixed in https://github.com/elastic/kibana/pull/144093/commits/5db5d415564d7a4a92441a06029b937c81fad194

jgowdyelastic reviewed (approved):
Added a few comments that can be looked at later, but on the whole code LGTM

darnautov commented on x-pack/plugins/aiops/public/components/change_point_detection/chart_component.tsx:
```diff
@@ -0,0 +1,214 @@
+    [dataView.id, requestParams.splitField, annotation.group_field]
+  );
+
+  // @ts-ignore
```
added in https://github.com/elastic/kibana/pull/144093/commits/d1f083ee7771a4bb386a230c3ebaa75fc5f1dcc0
